sdk/servicebus/azservicebus/sender.go

```diff
@@ -16,31 +18,86 @@ type Sender struct {
 		queueOrTopic string
 	}
 
-	createSender func(ctx context.Context) (legacySender, error)
-
 	mu           *sync.Mutex
-	legacySender legacySender
+	legacySender internal.LegacySender
 
 	linkState *linkState
+
+	// for testing
+	nsCreateLegacySender func(ctx context.Context, entityPath string) (internal.LegacySender, error)
+}
+
+// SendableMessage are sendable using Sender.SendMessage.
+// Message, MessageBatch implement this interface.
+type SendableMessage interface {
+	toAMQPMessage() (*amqp.Message, error)
+}
+
+type messageBatchOptions struct {
+	maxSizeInBytes *int
+}
+
+// MessageBatchOption is an option for configuring batch creation in
+// `NewMessageBatch`.
+type MessageBatchOption func(options *messageBatchOptions) error
+
+// MessageBatchWithMaxSize overrides the max size (in bytes) for a batch.
+// By default NewMessageBatch will use the max message size provided by the service.
+func MessageBatchWithMaxSize(maxSizeInBytes int) func(options *messageBatchOptions) error {
+	return func(options *messageBatchOptions) error {
+		options.maxSizeInBytes = &maxSizeInBytes
+		return nil
+	}
+}
+
+// NewMessageBatch can be used to create a batch that contain multiple
+// messages. Sending a batch of messages is more efficient than sending the
+// messages one at a time.
+func (s *Sender) NewMessageBatch(ctx context.Context, options ...MessageBatchOption) (*MessageBatch, error) {
```

Comment:
Conceptually, it seems weird that creating a message batch (which should just be a container of messages) needs a context.  Can we defer creating the sender until we are actually attempting to send?

Reply:
It is a bit odd, but all the SDKs do this the same way.

The gist is that we need to create a link before we know the actual max message size (it comes with the initial attaching of the link itself). Without this the batching feature isn't useful since it can't properly check that the messages you're adding to the batch are properly sized at all.

So this is one case where we can't defer it.